core/src/main/java/com/linecorp/armeria/server/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,171 @@
+        final ChannelFuture close = channel.close();
+        close.addListener(future -> {
+            if (future.isSuccess()) {
+                logger.debug("Closed channel: {} as PING timed out.", channel);
+            } else {
+                logger.debug("Cannot close channel: {}.", channel, future.cause());
+            }
+        });
+    };
+
+    private final GenericFutureListener<ChannelFuture> pingWriteListener = future -> {
+        final EventLoop el = future.channel().eventLoop();
+        if (future.isSuccess()) {
+            shutdownFuture = el.schedule(shutdownRunnable, pingTimeoutInNanos, TimeUnit.NANOSECONDS);
+            state = State.PENDING_PING_ACK;
+            stopwatch.reset().start();
+        } else {
+            // write failed, likely the channel is closed.
+            logger.debug("PING write failed for channel: {}", channel);
+        }
+    };
+
+    Http2KeepAliveHandler(Channel channel, Http2FrameWriter frameWriter, long pingTimeoutInNanos) {
+        this.channel = channel;
+        this.frameWriter = frameWriter;
+        this.pingTimeoutInNanos = pingTimeoutInNanos;
+    }
+
+    public void onChannelIdle(ChannelHandlerContext ctx, IdleStateEvent event) {
+        checkState(state == State.IDLE, "Waiting for PING ACK or shutdown");
+
+        // Only interested in ALL_IDLE event.
+        if (event.state() != IdleState.ALL_IDLE) {
+            return;
+        }
+
+        logger.debug("{} event triggered on channel: {}. Sending PING", event, channel);
+        writePing(ctx);
+    }
+
+    private void writePing(ChannelHandlerContext ctx) {
+        lastPingPayload = random.nextLong();
+        pingWriteFuture = frameWriter.writePing(ctx, false, lastPingPayload, ctx.newPromise())
+                                     .addListener(pingWriteListener);
+        state = State.PING_SCHEDULED;
+        ctx.flush();
+    }
+
+    public void onChannelInactive() {
+        state = State.SHUTDOWN;
+        if (shutdownFuture != null) {
+            shutdownFuture.cancel(false);
+            shutdownFuture = null;
+        }
+        if (pingWriteFuture != null) {
+            pingWriteFuture.cancel(false);
+            pingWriteFuture = null;
+        }
+    }
+
+    public void onPingAck(long data) throws Http2Exception {
+        final long elapsed = stopwatch.elapsed(TimeUnit.NANOSECONDS);
+
+        if (state != State.PENDING_PING_ACK) {
+            throw new Http2Exception(Http2Error.PROTOCOL_ERROR,
+                                     "State expected PENDING_PING_ACK but is " + state);
+        }
+        if (lastPingPayload != data) {
+            throw new Http2Exception(Http2Error.PROTOCOL_ERROR,
+                                     "PING received but payload does not match. " + "Expected: " +
+                                     lastPingPayload + ' ' + "Received :" + data);
+        }
+        if (shutdownFuture != null) {
+            shutdownFuture.cancel(false);
```

Comment:
Is it ok to cancel, or should I have a check in shutdownRunnable to check if state is IDLE and then do **not** close the channel

Reply:
- How about checking the return value of `cancel()` and log like 'Received a late PING(ACK=1) in {}ns?
- I think it's better cancelling here.